File: core/Transaction.ts
export enum TransactionType {
  DEPOSIT = "DEPOSIT",
  WITHDRAWAL = "WITHDRAWAL",
}

export class Transaction {
  constructor(
    public readonly type: TransactionType,
    public readonly amount: number,
  ) {}

  static deposit(amount: number) {
    return new this(TransactionType.DEPOSIT, amount);
  }

  static withdraw(amount: number) {
    return new this(TransactionType.WITHDRAWAL, amount);
  }
}

export interface Transactional {
  deposit(amount: number): Transaction;
  withdraw(amount: number): Transaction;
}

File: core/transaction.py
from enum import Enum

class TransactionType(Enum):
    DEPOSIT = 1
    WITHDRAWAL = 2

class Transaction:
    def __init__(self, type, amount):
        self.type = type
        self.amount = amount

    @classmethod
    def new(cls, type, amount):
        return cls(type, amount)

    @classmethod
    def deposit(cls, amount):
        return cls(TransactionType.DEPOSIT, amount)

    @classmethod
    def withdraw(cls, amount):
        return cls(TransactionType.WITHDRAWAL, amount)
